### PR TITLE
Docker shim INFRA-2

### PR DESCRIPTION
**Pull request by jdmulloy**
Configuration Pull Request
---

Make sure that the following steps are done before merging:

  - [ ] A DevOps team member has approved the PR if it is code shared across multiple services and you don't own all of the services.
  - [ ] Are you adding any new default values that need to be overridden when this change goes live? If so:
    - [ ] Update the appropriate internal repo (be sure to update for all our environments)
    - [ ] If you are updating a secure value rather than an internal one, file a DEVOPS ticket with details.
    - [ ] Add an entry to the CHANGELOG.
  - [ ] If you are making a complicated change, have you performed the proper testing specified on the [Ops Ansible Testing Checklist](https://openedx.atlassian.net/wiki/display/EdxOps/Ops+Ansible+Testing+Checklist)?  Adding a new variable does not require the full list (although testing on a sandbox is a great idea to ensure it links with your downstream code changes).
  - [ ] Think about how this change will affect Open edX operators.  Have you updated the wiki page for the next Open edX release?

### CODE REVIEW

File: playbooks/roles/enterprise_catalog/meta/main.yml [hunk 1]
```diff
@@ -12,6 +12,7 @@
 
 dependencies:
   - role: edx_django_service
+    EDX_DJANGO_SERVICE_ENABLE_DOCKER_SHIM: true
```
Comment (bderusha) [inline playbooks/roles/enterprise_catalog/meta/main.yml]: NOTE: We should not merge this with this flag set to `true` without consulting the team that own it.

File: playbooks/roles/edx_django_service/tasks/main.yml [hunk 2]
```diff
@@ -239,12 +263,14 @@
...
     dest: "{{ COMMON_BIN_DIR }}/{{ item.split('.')[0] }}.{{ edx_django_service_name }}"
     state: link
+  when: not EDX_DJANGO_SERVICE_ENABLE_DOCKER_SHIM
   with_items:
     - manage.py
   tags:
     - install
     - install:app-requirements
 
+# TODO: How to handle this in docker build?
```
Comment (bderusha) [inline playbooks/roles/edx_django_service/tasks/main.yml]: I agree with @coryleeio 's comment made out-of-band 

>  run the make command inside the container with exec when the flag is enabled

File: playbooks/roles/edx_django_service/defaults/main.yml [hunk 3]
```diff
@@ -132,3 +132,5 @@ edx_django_service_migration_check_services: "{{ edx_django_service_name }}"
 EDX_DJANGO_SERVICE_ENABLE_ADMIN_URLS_RESTRICTION: false
 
 EDX_DJANGO_SERVICE_ADMIN_URLS: []
+
+EDX_DJANGO_SERVICE_ENABLE_DOCKER_SHIM: false
```
Comment (bderusha) [inline playbooks/roles/edx_django_service/defaults/main.yml]: I don't love the term shim here.  Running a docker container on EC2 is a valid approach.  It's a shim for us as we target k8s, but I would rather drop it from the variable name

Comment (adzuci) [reply playbooks/roles/edx_django_service/defaults/main.yml]: Maybe `ENABLE_SINGLE_CONTAINER_MODE`?

Even though it's a valid approach it isn't widely suggested and I do like the idea of discouraging it's use.
https://docs.docker.com/get-started/orchestration
image

File: playbooks/roles/edx_django_service/templates/edx/app/app/app.sh.j2 [hunk 4]
```diff
@@ -2,6 +2,10 @@
 
 # {{ ansible_managed }}
 
+{% if EDX_DJANGO_SERVICE_ENABLE_DOCKER_SHIM %}
+cd ..
+docker-compose up
```
Comment (coryleeio) [inline playbooks/roles/edx_django_service/templates/edx/app/app/app.sh.j2]: may need to exec here

File: playbooks/roles/edx_django_service/meta/main.yml [hunk 5]
```diff
@@ -1,6 +1,10 @@
 ---
 dependencies:
   - common
+  - role: docker-tools
+    docker_users:
+      - '{{ edx_django_service_automated_users }}'
+      - "{{ edx_django_service_user }}"
   - role: supervisor
     supervisor_spec:
       - service: "{{ edx_django_service_name }}"
```
Comment (coryleeio) [inline playbooks/roles/edx_django_service/meta/main.yml:10]: I note that just below this it refers to python.service_name, is that still needed?

File: playbooks/roles/edx_django_service/tasks/main.yml [hunk 6]
```diff
@@ -21,19 +21,39 @@
...
+  command: "id -u {{ edx_django_service_user }}"
+  register: edx_django_service_user_id
+  when: edx_django_service_enable_experimental_docker_shim
+  tags:
+    - install
+    - install:system-requirements
+
+- name: Docker build
+  when: edx_django_service_enable_experimental_docker_shim
+  command: make docker_build
```
Comment (coryleeio) [inline playbooks/roles/edx_django_service/tasks/main.yml]: I wonder if we should force them to implement this in a make file or just call "docker build ."

File: playbooks/roles/edx_django_service/defaults/main.yml [hunk 7]
```diff
@@ -263,3 +265,8 @@ edx_django_service_celery_heartbeat_enabled: true
 edx_django_service_lang: 'en_US.UTF-8'
 edx_django_service_default_stopwaitsecs: 432000
 edx_django_service_workers_supervisor_conf: '{{ edx_django_service_name }}-workers.conf'
+
+edx_django_service_enable_experimental_docker_shim: false
+# Default to blank if edx_django_service_enable_experimental_docker_shim is false
+edx_django_service_docker_run: ""
```
Comment (coryleeio) [inline playbooks/roles/edx_django_service/defaults/main.yml:271]: run_command might be more clear

**Review by coryleeio — APPROVED** [review]

File: playbooks/roles/edx_django_service/tasks/main.yml (hunk 6 above)
Comment (jdmulloy) [reply playbooks/roles/edx_django_service/tasks/main.yml]: @coryleeio Good thought. The playbook already assumes ```make static``` and ```make migrations```. This is ```make docker_build``` for enterprise catalog. Should I hard code this in Ansible or should I let the app decide what it should do in the Makefile? The playbook does assume that an image named :latest and :latest-newrelic gets built. I suppose I could just build the one we need based on if new relic is enabled or not.

**Review by coryleeio — APPROVED** [review]

**Pull request merged**